File: javascript/leetcode/rotate-function.js
/**
 * @param {number[]} A
 * @return {number}
 */
var maxRotateFunction = function(A) {
    if (A.length == 0) { return 0; }
    var rSum = 0;
    var allSum = 0;
    for (var i = 0; i < A.length; i++) {
        var n = A[i];
        rSum += n * i;
        allSum += n;
    }
    var result = rSum;
    for (var i = A.length - 1; i >= 0; i--) {
        rSum = rSum + allSum - A.length * A[i];
        result = Math.max(rSum, result);
    }
    return result;
};

File: java/leetcode/rotate-function.java
class Solution {
    public int maxRotateFunction(int[] A) {
        if (A.length == 0) { return 0; }
        int rSum = 0;
        int allSum = 0;
        int n = 0;
        for (int i = 0; i < A.length; i++) {
            n = A[i];
            rSum += n * i;
            allSum += n;
        }
        int result = rSum;
        for (int i = A.length - 1; i >= 0; i--) {
            rSum = rSum + allSum - A.length * A[i];
            result = Math.max(rSum, result);
        }
        return result;
    }
}
